### PR TITLE
New Label Engine

### PR DESCRIPTION
**Pull request by cram42**
# Description

Full custom PDF label support. Allows single page labels (for tape printers) or rectangles-on-a-page sheets (for label sticker paper). Renders to PDF using TCPDF.
Labels are defined as class files within `App\Models\Labels`, extending the `Label `abstract class. These define unit of measure, size, and feature support.
Helper functions are in place to make placing text, images, and barcodes easy.
Helper abstract classes are in place to allow multiple labels per page.
API endpoints have been created to find Label templates installed.
A custom (read: hacky) label `DefaultLabel` has been created to simulate the legacy label generator. When selected, all legacy settings remain available.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Tested in my browser, as I don't know how to run the included test units (missing testing database).

**Test Configuration**:
* PHP version: 7.4.30
* MySQL version: 10.5.15-MariaDB
* Webserver version: Apache/2.4.54 (Debian)
* OS version: Debian GNU/Linux 11 (bullseye)

# Checklist:

- [X] I have read the Contributing documentation available here: https://snipe-it.readme.io/docs/contributing-overview
- [X] I have formatted this PR according to the project guidelines: https://snipe-it.readme.io/docs/contributing-overview#pull-request-guidelines
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by snipe** [comment]
Hi there - thanks for this! This is a pretty big PR - can you include some screenshots, etc?

**Comment by snipe** [comment]
My inclination for revamping labels was actually going to be javascript-based, not PDF based, so people could change things on the fly and see how they'd look. That said, I'm curious to check out what you've done here.

**Comment by cram42** [comment]
Hi!

Yeah, I tried a few options with JS and CSS but found the results when printing were inconsistent between printers and even between Chrome's print page and the system dialog. That's why I went with PDF, as we can set the exact page size and it will always print just so. That means it worked with label sheets (I tested on Avery L7162), and also on tape labels (Brother P-Touch tze 24mm).

Your idea of an instant preview is awesome. I'll see if I can get something like that working somehow.

I'll get some screenshots and photos together and attach them here.

**Comment by cram42** [comment]
Each example below shows the Labels settings page, the assets included, the output from clicking **Generate Label**, and the output from selecting multiple on the assets index page and choosing **Generate Labels** from the bulk actions.

### New Engine Disabled - Settings Page
Settings - Disabled

### New Engine Enabled - DefaultLabel Selected
Settings - Enabled - DefaultLabel

### New Engine Enabled - Custom Template Selected
Settings - Enabled - Custom Template

### Test Prints
TestPrints

**Comment by cram42** [comment]
### New example using 24mm P-touch label tape
P-touch Labels

**Comment by cram42** [comment]
The option for **2D Barcode Target** allows the QR code URL to include the asset tag so we can use a portable scanner for quick identification. The new /ht/{asset_tag} route is a shorter version of /hardware/byTag/{asset_tag}, to allow for smaller QR codes.

**Comment by brianhoganm** [comment]
Looks good. But, can we add specific common fields to the label. *Company*,
*Location*, *Asset Tag, Serial* and *Person (First & Last Name)* (IF ITEM
IS CHECKED OUT). This would be very useful for creating student chromebook
labels.

**Comment by cram42** [comment]
> Looks good. But, can we add specific common fields to the label. *Company*, *Location*, *Asset Tag, Serial* and *Person (First & Last Name)* (IF ITEM IS CHECKED OUT). This would be very useful for creating student chromebook labels.

**Edit:** Just realised you wanted First Name and Last Name. Currently, they're separate fields, but I'll have a look at building a definition for concatenation.
**Edit 2:** OK, simple, there's a fullname attribute we can use. So `Assigned To=assignedto.fullname;` works.

Sure can. Currently, it's not super user-friendly. But, I just set the **Fields Definition** to `Company=company.name;Asset Tag=asset_tag;Serial=serial;Assigned To=assignedto.username;`, and removed any titles. Here's the result:

Fields

**Comment by cram42** [comment]
### P-touch 12mm Example
P-touch 12mm

**Comment by cram42** [comment]
Can now test label layouts by name, with test data.

**Comment by cram42** [comment]
Preview feature created. Instead of having to save settings and refresh another tab. You can now just hit "Refresh Preview" and the current setting form values are passed directly to the URL params in the iFrame.

**Comment by cram42** [comment]
Added Alpine.js based helper control to make setting field definitions more user-friendly.
As far as the form is concerned, it just sees a `<input type="hidden">`.

**Comment by cram42** [comment]
Potentially solves issues:
#9253 #9425 #9458 #9851 #9964 #10284 #10557 #10643 #11360 #11095 #11747 #11961 #11966

File: app/Http/Controllers/Api/LabelsController.php
Comment (snipe) [inline app/Http/Controllers/Api/LabelsController.php:23]: I don’t see this permission actually being created or settable anywhere in the code?

File: app/Http/Controllers/Assets/AssetsController.php
Comment (snipe) [inline app/Http/Controllers/Assets/AssetsController.php]: TINY nit here (and I’d accept this without this change) but stylistically, we typically order method parameters with `Request` first, required (not nullable) variables next, and nullable/optional variables last.

File: app/Http/Controllers/LabelsController.php
Comment (snipe) [inline app/Http/Controllers/LabelsController.php]: Again, TINY nit - can we rename this to something a little clearer, like `$exampleAsset` or `$previewAsset`?

File: composer.json
Comment (snipe) [inline composer.json:73]: Is there a reason we need to add a second PDF library here? IIRC, we already include one:

https://github.com/snipe/snipe-it/blob/master/composer.json#L30

File: routes/web/hardware.php
Comment (snipe) [inline routes/web/hardware.php:191]: What’s `ht` here? Can you clarify what this new route does?

**Review by snipe — CHANGES_REQUESTED** [review]
Hey there - sorry this took us a bit to get to. PRs with 47 changed files are a little daunting.

I have some questions and clarifications inline, and a few small nits. 

This PR looks gorgeous though. Really nice work!

I admit I’m a little loathe to add Alpine.js here - we’ve avoided it up until now, and we already carry the weight of jquery (which is required for the admin theme), vue.js (for now, though we are trying to phase it out) and Livewire.  Adding another JS framework kinda gives me the oogs.

What’s the difference in page weight? Also, I’m not that familiar with Alpine - is the `defer.js` part of that? It’s very long, so I’m not 100% sure what it’s needed for. It’s not necessarily a blocker, but we’re trying to pare down the JS dependencies we have, so adding more gives me a little bit of the oogs.

**Comment by snipe** [comment]
Hey there - just reaching back out on this one. I had a few questions (and now we've got a few conflicts). This looks pretty cool and we're definitely interested in taking it!

File: routes/web/hardware.php
Comment (MoralCode) [reply routes/web/hardware.php:191]: maybe this isnt helpful as Im not the OP, but a [previous comment](https://github.com/snipe/snipe-it/pull/12050#issuecomment-1299471546) in the PR mentioned that `ht` is a shorter version of the existing url endpoint to allow the QR code data to be shorter/smaller/simpler.

**Comment by MoralCode** [comment]
just noticed @iangozer's related pull requests and am curious if theres a middleground here to accomodate people with both sheet and single stream style label printers. I've done some work in the past with a project (unrelated to asset management) that used SVG files with placeholder values as templates for generating labels. Being unfamiliar with this code that's probably a major change to make, but it might be a way to also tackle the "extra pdf dependency" by moving the need for PDFs out of the workflow for the single-stream-printer folks. 

~~It may also be a cool opportunity to somehow integrate with the various [template codes](https://www.avery.com/templates) that label manufacturers (like avery) seem to provide for print on the top and bottom edge of every sheet of labels (and provide corresponding word/photoshop .etc templates for). Maybe these can be used to automate some aspects of the layout and spacing for these sheet style labels so all the user has to do is say "hey im using 5162 labels" (or the [L7162 ones from earlier](https://github.com/snipe/snipe-it/pull/12050#issuecomment-1299405943), which seem to be Australian specific... hmm maybe thats a wrinkle in this idea) and pick/make a template and theyre good to go.~~  
Edit: it helps if I actually read the contents of the pull request to see that this was already included

Hope this is helpful

**Comment by iangozer** [comment]
> just noticed @iangozer's related pull requests and am curious if theres a middleground here to accomodate people with both sheet and single stream style label printers. I've done some work in the past with a project (unrelated to asset management) that used SVG files with placeholder values as templates for generating labels. Being unfamiliar with this code that's probably a major change to make, but it might be a way to also tackle the "extra pdf dependency" by moving the need for PDFs out of the workflow for the single-stream-printer folks.
> 
> It may also be a cool opportunity to somehow integrate with the various [template codes](https://www.avery.com/templates) that label manufacturers (like avery) seem to provide for print on the top and bottom edge of every sheet of labels (and provide corresponding word/photoshop .etc templates for). Maybe these can be used to automate some aspects of the layout and spacing for these sheet style labels so all the user has to do is say "hey im using 5162 labels" (or the [L7162 ones from earlier](https://github.com/snipe/snipe-it/pull/12050#issuecomment-1299405943), which seem to be Australian specific... hmm maybe thats a wrinkle in this idea) and pick/make a template and theyre good to go.
> 
> Hope this is helpful

The changes I am making will still work fine for sheets, and no reason presets couldn't be added.
There are two key things to the pdf solution though:
   1. PDFs prevent the different ways browsers handle printing html including margins etc.
   2. PDFs add an extra step or to when printing, fine if you're doing sheets, a real pain in the but if you're doing singles.

Keen to make some progress but waiting on some answers to how things are currently handled.

File: app/Http/Controllers/LabelsController.php
Comment (MoralCode) [reply app/Http/Controllers/LabelsController.php]: i have a branch on my fork that renames this to `$exampleAsset` in case that's useful in helping get this merged

File: app/Http/Controllers/Assets/AssetsController.php
Comment (MoralCode) [reply app/Http/Controllers/Assets/AssetsController.php]: in case its useful, i swapped these parameters on my fork since it was a relatively easy, low risk change to make. I don't have the right access to directly add it here, but my branch is based on this one so it should be easy for someone with write access to push the commits to the right place

File: app/Http/Controllers/Api/LabelsController.php
Comment (MoralCode) [reply app/Http/Controllers/Api/LabelsController.php:23]: Is having a permission for this feature something that would be useful to include in this PR? if its not doing anything, it seems like its pretty easy to just remove unless there's a need/desire to have it behind a permission setting

**Comment by MoralCode** [comment]
> PDFs prevent the different ways browsers handle printing html including margins etc.

Can you elaborate on this a little more? I think @cram42 initially tried a non-PDF solution but couldn't get it to behave consistently across many browsers and printers as described in https://github.com/snipe/snipe-it/pull/12050#issuecomment-1299387699. It seems like there may be reasons both for and against allowing browsers to handle printing themselves

File: app/Models/Labels/Sheet.php
Comment (MoralCode) [inline app/Models/Labels/Sheet.php:5]: To me it doesn't seem especially intuitive that a sheet is an extension of a single label. To me it feels like composition may be a cleaner solution here by making `Sheet` is its own standalone class that represents a composition of many  `Label` objects

File: app/Http/Controllers/Api/LabelsController.php
Comment (cram42) [reply app/Http/Controllers/Api/LabelsController.php:23]: > Adding another JS framework kinda gives me the oogs

Yeah, I get it! I really didn't want to add another one. I tried using Livewire, but it made me sad. I just can't get me head around it to do what I want.

> What’s the difference in page weight

93.8k :(

> is the defer.js part of that?

Sadly it is. Alpine's loading has to be after the DOM. So I needed to mix an "all-defer.js" and load it as <script defer src=...


I'll revisit a Livewire option now that I'm back from holidays.

Comment (cram42) [reply app/Http/Controllers/Api/LabelsController.php:23]: > I don’t see this permission actually being created or settable anywhere in the code?

Oops, nope. It was really just a placeholder in case locking it down was required.

> Is having a permission for this feature something that would be useful to include in this PR? if its not doing anything, it seems like its pretty easy to just remove unless there's a need/desire to have it behind a permission setting

I can remove it for now and reintroduce if needed.

File: app/Http/Controllers/Assets/AssetsController.php
Comment (cram42) [reply app/Http/Controllers/Assets/AssetsController.php]: Yep. This was a derp on my part. Sorry.

File: composer.json
Comment (cram42) [reply composer.json:73]: The included one (dompdf) requires building from HTML DOM which defeats the purpose of this engine.
TCPDF seems to be the most up-to-date library that lets you build from scratch. Also, tc-lib-barcode from the same author is already included and integrates nicely.

File: routes/web/hardware.php
Comment (cram42) [reply routes/web/hardware.php:191]: "ht" stands for Hardware Tag. The route is equivalent to [baseurl]/hardware/bytag/[hardwaretag].
It was created so that URLs can be shorter, and therefore QR codes can be simpler.

File: app/Models/Labels/Sheet.php
Comment (cram42) [reply app/Models/Labels/Sheet.php:5]: I actually wrote it that way in my initial testing, then found I was repeating the same code again.
I've made it this way so theoretically any label can be included as a template inside any other label.

Comment (MoralCode) [reply app/Models/Labels/Sheet.php:5]: what code was being repeated?

Maybe I'm too new to this codebase or thinking too high level to see why this didn't work, but couldn't you have the Sheet define the margins and spacing for the sheet (like the word templates do with the tables in the Avery template word docs) and then call a function in each label within the sheet to have it render itself and plonk that into its table square in the sheet to build the whole sheet?

File: composer.json
Comment (MoralCode) [reply composer.json:73]: > which defeats the purpose of this engine.

given previous comments in this PR it seems to me like the purpose of the engine (or at least using PDF) is to make it easier to align labels for users with sheet-style labels instead of the single label printers. Is that right? 
I think this was discussed a bit in the non-review comments, but it seems like the extra PDF step is cumbersome for users who  have existing workflows that are using single-label printers.

What issues were you having with the existing HTML label layouts that prompted you to make this? would other solutions such as CSS (or even print specific stylesheets), SVG labels, or something else that's closer to HTML help eliminate the need for a new PDF library and make things both 1) as to use as the current method for single label people, and 2) consistently lay out labels in sheets for people who print labels on sheets?

is it possible to craft an HTML dom input to dompdf that causes it to generate PDF's with a certain layout? then it would simplify the problem from "write HTML that perfectly lays out a sheet of labels on all browsers/devices/print methods" to "write HTML that perfectly lays out a sheet of labels specifically for dompdf". Then TCPDF could be eliminated and also the PDFs could be used only by people printing sheets

**Comment by Kaphalor** [comment]
Hey guys,
any information or ETA when this will be available on hosted instances?
Would make our lives way more easy regarding labels and metric conversions!

**Comment by MoralCode** [comment]
> any information or ETA when this will be available on hosted instances?
> Would make our lives way more easy regarding labels and metric conversions!

I'm not the maintainers but i have been working on my own external tools for both [generating labels in bulk](https://github.com/MoralCode/asset-labelmaker) as PNG images and (most recently) throwing together a [webpage](https://github.com/MoralCode/LabelSheetGenerator) where you can select these PNGs and have it lay them out on a sheet of labels for you. These tools were made as  pretty low-fi/hacky solutions, but may help provide a bridge for anyone with a similar usecase to me while waiting for this feature to land

File: composer.json
Comment (cram42) [reply composer.json:73]: The purpose is to make it easier for sheet-style, tape-style, and single label templates to be repeatable across browsers and devices.

I started out attempting to use 12mm and 24mm PTouch tapes for asset tags, but found it impossible to print tags in a way that fit correctly. Each browser would decide to add extra margins, or flip labels 90 degrees when batch printing, or not trigger auto-cutting for each page.

Then I tried setting up for Avery label sheets instead. Got it to fit my sheet after about an hour of converting mm from a ruler to hundredths of inches and trial-and-error. Then I realised after all that, it's kinda ugly (no offense to the default setup).

Thought "if only there was a format for printing instead of display". Hence PDF. Add support for multiple layouts and fully customisable fields to show and bam, got a stupidly large chunk of code.

**Review by cram42 — COMMENTED** [review]
Also, this removes the unused authorization stuff.

**Comment by snipe** [comment]
@uberbrady - when you get a moment, can you help resolve these conflicts? (@cram42 has already done a lot here, and it's our fault for not being faster on the draw.) 

Once we have, I'd love to us to sneak this into v6.1. Even if the specific labels need some extra work, the fallback scenario (back to legacy) is pretty solid IMHO.

**Comment by Godmartinz** [comment]
Hi @cram42 This looks really good! A few of us would like to test it and hopefully get this merged in sooner than later. But that means I have to ask, can you resolve the merge conflicts? Some things have changed since you opened the PR (a loooong time ago), I'm hoping its not a huge undertaking for you. 

Looking forward to testing this out 😁

**Comment by SonGokussj4** [comment]
This would help so much right now because my tomorrow job is to get the service tag to the labels and I just found this thread.

Is there any estimation? Day/Week/2 months? This looks so so so good.

**Comment by cram42** [comment]
> Hi @cram42 This looks really good! A few of us would like to test it and hopefully get this merged in sooner than later. But that means I have to ask, can you resolve the merge conflicts? Some things have changed since you opened the PR (a loooong time ago), I'm hoping its not a huge undertaking for you.
> 
> Looking forward to testing this out 😁

Woo, yeah! So much has changed and I've been snowed under for months!
Working on resolving the conflicts (hoping this week will get it sorted).

**Comment by cram42** [comment]
> @uberbrady - when you get a moment, can you help resolve these conflicts? (@cram42 has already done a lot here, and it's our fault for not being faster on the draw.)
> 
> Once we have, I'd love to us to sneak this into v6.1. Even if the specific labels need some extra work, the fallback scenario (back to legacy) is pretty solid IMHO.

Sorry I didn't get a chance to try get rid of the extra dependencies yet. Just got to fix the conflicts in the changes to the labels blade, making sure I'm not breaking stuff.

**Comment by cram42** [comment]
Resolved conflicts.

**Comment by MagicLike** [comment]
I am currently not able to go over all the details of this PR, but does it implement https://github.com/snipe/snipe-it/issues/10557#issuecomment-1184343114 in some way?

**Comment by cram42** [comment]
> I am currently not able to go over all the details of this PR, but does it implement [#10557 (comment)](https://github.com/snipe/snipe-it/issues/10557#issuecomment-1184343114) in some way?

Any field can be added, as long as you know its internal id.

image

**Comment by MagicLike** [comment]
> > I am currently not able to go over all the details of this PR, but does it implement [#10557 (comment)](https://github.com/snipe/snipe-it/issues/10557#issuecomment-1184343114) in some way?
> 
> Any field can be added, as long as you know its internal id.
> 
> image
> 

Nice!

**Comment by snipe** [comment]
This kind blew up Codacy for code quality :-/

We probably want to refactor this a bit. 
https://app.codacy.com/gh/snipe/snipe-it/commit?cid=754883792&bid=3362446

**Comment by snipe** [comment]
The show full screen icon doesn't seem to do anything?

Screenshot 2023-08-15 at 6 33 59 PM

It just opens a new blank window in safari.

Also we should probably be consistent with our placeholders/variables, at least as far as the users are concerned. We use `{FOO}` elsewhere in the system, like for the warranty URL.

Screenshot 2023-08-15 at 6 35 16 PM

I'm honestly finding this UI a little confusing and I'm not sure how it's supposed to work. 

I'm getting `[2023-08-15 10:41:19] local.ERROR: Error: Class 'TCPDF_STATIC' not found in /Users/snipe/Sites/snipe-it/snipe-it/app/Models/Labels/Label.php:537` when I try to load the list

**Comment by snipe** [comment]
I might have to back this change out at least until we handle these issues :(

**Comment by snipe** [comment]
So I think the reason the preview full screen page is blank is because I can't get the list to load, so it can't preview anything. :-/. I think we need to figure out the TCFPDF_STATIC issue first.

**Comment by snipe** [comment]
Okay, I've fixed the composer conflicts and issues. Still seeing an issue with the rendering if assets are selected that don't belong to a company, since the method in `Views/Label.php` assumes `$asset->company` exists and then tries to grab the name property. :( Lots of people don't use companies with their assets, so it 500s out with "Trying to get property on non-object when trying to get the name of as non-existent property). Trying to fix this now. 

I'm also not really in love with creating a new directory (`app\View`) when nothing else lives in there. I might refactor that as well.

**Comment by snipe** [comment]
Related subsequent PRs:

- #13465
- #13466
- #13467
- #13469
- #13473
- #13489
- #13491
- #13492

**Comment by snipe** [comment]
Also possibly related, though it might be an asset tag compatibility issue: #13472

**Comment by snipe** [comment]
A few additional things.... I have 30 set as my "per page", but only 20 are showing up. 

Screenshot 2023-08-16 at 1 37 17 AM

And I'm not 100% convinced that this field is even necessary?

Screenshot 2023-08-16 at 1 36 18 AM

I think it's probably more like "use a logo", since I'd expect Snipe-IT's label logo to be the default, unless the company it belongs to (if it has one) has a logo.

**Comment by snipe** [comment]
@cram42 this may be a very stupid question, but I'm not sure how to add a new label template. I added a directory within the `Tapes` directory called `Dymo`, but it's not picking it up in the label template listings. Seems like it's just trying to scan that directory, but it's not showing :-/

**Comment by snipe** [comment]
Never mind, I think I've got it

**Comment by najirod** [comment]
@snipe 
I think there is a problem with small labels, qr code does not scale or something..

Screenshot 2023-08-28 at 12 23 51 PM
Screenshot 2023-08-28 at 12 23 20 PM
Screenshot 2023-08-28 at 12 21 45 PM

**Comment by snipe** [comment]
@cram42 Are you able to take a look at @najirod's issue, or should we try to handle it?

**Comment by Robert-Azelis** [comment]
I have a missing some options which would be useful for use, like:
- align per field (right, centre, left)
- align logo (left, right)
- add/edit/delete templates (including rename)
- selection of template to use before print labels

**Comment by snipe** [comment]
> add/edit/delete templates (including rename)

Renaming wouldn't make sense, since they are specific  to the labels themselves.

> selection of template to use before print labels

Not sure I understand. You can already select the template before printing. 

> DataSource: drop-down list of available attributes to use

Probably not feasible, as there are a ton of fields across assets, models, categories, etc. Would be a UI nightmare.

**Comment by Robert-Azelis** [comment]
> > selection of template to use before print labels
> 
> Not sure I understand. You can already select the template before printing.
> 
If you mean about select template in Labels setting, please bear in mind not all users have access, only admin, so if some user would like to generate labels from assets is not able select template to use (shets/tapes)

**Comment by Capobuf** [comment]
> > > I am currently not able to go over all the details of this PR, but does it implement [#10557 (comment)](https://github.com/snipe/snipe-it/issues/10557#issuecomment-1184343114) in some way?
> > 
> > 
> > Any field can be added, as long as you know its internal id.
> > image
> 
> Nice!

First of all, thanks for the great job!

I'm a simple user but...how can we find the internal id from, for example, a custom field?

Thanks a lot!

**Comment by asuman1179** [comment]
I am trying to create my own Labels\Tapes\Brother setup but I am having issues trying to understand how to change the const to match what I need.  @cram42  I am remarkably close but for some reason I have a larger space at the bottom of the tag that I just can figure out where I have gone wrong.  The Tape i am using is DK-2211 and i selected 29mm x 45mm as the size for the setup.
tempsnip

**Comment by asuman1179** [comment]
Figured it out. I had removed the tag from below the QR code and put in the right column but some other place holder is holding that space.

**Comment by najirod** [comment]
> @cram42 Are you able to take a look at @najirod's issue, or should we try to handle it?

@snipe 
I think my issue about small labels is not fixed yet, I tried again to check but found that there is another bug when you select the default label and setup with barcode it gives "TCPDF ERROR: Error in 1D barcode string"

Screenshot 2023-09-27 at 9 55 20 AM

**Comment by asuman1179** [comment]
Figured it out. I had removed the tag from below the QR code and put in the right column but I guess some other place holder is holding that space that I had not figured out just yet.

**Comment by cram42** [comment]
Sorry all for the absence. My grandfather's had a stroke so I've been at the hospital all day every day.


> @snipe I think there is a problem with small labels, qr code does not scale or something..
> [snip]

Yep, I get the feeling there may be a minimum QR size. Will check it out now.


> [snip]
> First of all, thanks for the great job!
> I'm a simple user but...how can we find the internal id from, for example, a custom field?
> Thanks a lot!

Under custom fields, you can enable the "DB Field" column.
image


> > @cram42 Are you able to take a look at @najirod's issue, or should we try to handle it? 
> @snipe I think my issue about small labels is not fixed yet, I tried again to check but found that there is another bug when you select the default label and setup with barcode it gives "TCPDF ERROR: Error in 1D barcode string"
> [snip]

I think it's to do with the barcode content vs code format.

I'm going to spin up a dev instance here at the hospital and see if I can get some stuff sorted while gramps sleeps.

**Comment by KWWL-Ethan** [comment]
> I am trying to create my own Labels\Tapes\Brother setup but I am having issues trying to understand how to change the const to match what I need. @cram42 I am remarkably close but for some reason I have a larger space at the bottom of the tag that I just can figure out where I have gone wrong. The Tape i am using is DK-2211 and i selected 29mm x 45mm as the size for the setup. tempsnip



> Figured it out. I had removed the tag from below the QR code and put in the right column but some other place holder is holding that space.

@asuman1179 can you share with the class how you created your own label template? It would be very nice to tweak the existing Brother 24mm template to include both a 1D and 2D barcode.

**Comment by Robert-Azelis** [comment]
I think will be great to add some option for create own template by "Save as" ;)
and very valuable if someone create simple instruction with description possible options how to adapt template.

**Comment by nikhil0426** [comment]
Hey there, sorry a bit new to Snipe-IT. Based on what I read above, if I know the internal ID, I can type it into the DataSource box under Field Definitions to make it appear on the label. I've tried doing this to add categories onto my label but regardless of whether I use "category" or "model.category" in the field definition, I can't get it to show up on my labels. Any advice?
image

**Comment by snipe** [comment]
You'd probably have to try `model.category.name`

**Comment by myriad007** [comment]
Try adding data source like: `_snipeit_project_name_6` Look at the posts above.

**Comment by bkgarry** [comment]
really silly question, should Assigned To be showing who it is Checked Out to? Ours have users checked out but showing nothing for assigned to when generating labels

**Pull request merged**
